[WFLY-10328] Agroal subsystem

barreiro opened the pull request:
- Issue https://issues.jboss.org/browse/WFLY-10328
- EAP feature https://issues.jboss.org/browse/EAP7-983
- Analisys https://github.com/wildfly/wildfly-proposals/pull/62

The current PR adds a subsystem that provides EE datasouces, based on the Agroal connection pool.
It's an alternative to the existing datasources subsystem that does not rely on the JCA container. It supports all of the EE requirements, including XA transactions, and has better performance characteristics.

---
Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [x] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue(s)
- [x] Pull Request contains description of the issue(s)
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [x] Analysis was done
- [x] Test Plan has been done
- [x] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

barreiro committed:
[WFLY-10328] Agroal subsystem

barreiro commented:
The source code for Agroal is in https://github.com/agroal/agroal
Agroal itself does not have any dependencies.

kabir commented:
@psakar @dosoudil https://github.com/agroal/agroal/blob/master/pom.xml#L30 says "Apache License Version 2.0" - is that still a problem? Apparently Agroal is already productised for Vert.x. License file at https://github.com/agroal/agroal/blob/master/README.md

jamezp commented:
This will require a rebase due to #11200

dmlloyd commented on agroal/src/main/java/org/wildfly/extension/agroal/definition/AgroalObjectAttribute.java:
```diff
@@ -0,0 +1,208 @@
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+public class AgroalObjectAttribute {
```
This (using customized parsers in a new subsystem) seems quite unusual.  Was there a discussion about this?

barreiro replied:
Yes there was. I wanted certain attributes to be grouped together in the model, that's why I went for ObjectAttribute, but that didn't marshal to the schema I had envisioned basically because ObjectAttribute parser/marshaller did't account for the 'group' of the attributes in it. Having a custom parser/marshaller is safer than trying to get this support upstream in WildFly core.

dmlloyd commented on agroal/src/main/java/org/wildfly/extension/agroal/deployment/DataSourceDefinitionInjectionSource.java:
```diff
@@ -0,0 +1,238 @@
+        ServiceBuilder<ManagedReferenceFactory> dataSourceServiceBuilder = phaseContext.getServiceTarget().addService( dataSourceServiceName, dataSourceService );
+
+        if ( transactional ) {
+            dataSourceServiceBuilder.addDependency( TxnServices.JBOSS_TXN_TRANSACTION_MANAGER, TransactionManager.class, dataSourceService.getTransactionManagerInjector() );
```
These services are deprecated.  For now it's appropriate to add a no-value dependency just on the ``JBOSS_TXN_TRANSACTION_MANAGER`` service, but otherwise just call ``ContextTransactionManager.getInstance()`` to get the transaction manager.  Similar for ``ContextTransactionSynchronizationRegistry``.

dmlloyd commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/DataSourceOperations.java:
```diff
@@ -0,0 +1,130 @@
+
+            // Define dependencies for JTA DataSources
+            if ( DataSourceDefinition.JTA_ATTRIBUTE.resolveModelAttribute( context, model ).asBoolean() ) {
+                serviceBuilder.addDependency( TxnServices.JBOSS_TXN_TRANSACTION_MANAGER, TransactionManager.class, dataSourceService.getTransactionManagerInjector() );
```
See above comment about JTA services.

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/DriverOperations.java:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        @Override
+        protected void populateModel(ModelNode operation, ModelNode model) throws OperationFailedException {
```
same as above (using the constructor with the attributes collection)

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/DriverOperations.java:
```diff
@@ -0,0 +1,152 @@
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String driverName = PathAddress.pathAddress( operation.require( OP_ADDR ) ).getLastElement().getValue();
```
`String driverName = context.getCurrentAddressValue()`

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/DriverOperations.java:
```diff
@@ -0,0 +1,152 @@
+    public static class DriverRemove extends AbstractRemoveStepHandler {
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String driverName = PathAddress.pathAddress( operation.require( OP_ADDR ) ).getLastElement().getValue();
```
`String driverName = context.getCurrentAddressValue()`

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java:
```diff
@@ -0,0 +1,122 @@
+
+    public static final String XADATASOURCE_SERVICE_NAME = "xa-datasource";
+
+    public static final ServiceName XADATASOURCE_SERVICE_PREFIX = ServiceName.of( "wildfly", "agroal", "xa-datasource" );
```
same question as above.

btw, you could create somewhere a `ServiceName AGROAL = ServiceName.of( "wildfly", "agroal")` and use it as the basis of all your other service names.

dmlloyd replied:
+1, shared base names can reduce memory usage.

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java:
```diff
@@ -0,0 +1,122 @@
+    private static class XADataSourceAdd extends AbstractAddStepHandler {
+
+        @Override
+        protected void populateModel(ModelNode operation, ModelNode model) throws OperationFailedException {
```
same remark as above

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java:
```diff
@@ -0,0 +1,122 @@
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String datasourceName = PathAddress.pathAddress( operation.require( OP_ADDR ) ).getLastElement().getValue();
```
`context.getCurrentAddressValue()`

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java:
```diff
@@ -0,0 +1,122 @@
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String datasourceName = PathAddress.pathAddress( operation.require( OP_ADDR ) ).getLastElement().getValue();
```
`context.getCurrentAddressValue()`

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java:
```diff
@@ -0,0 +1,122 @@
+            serviceBuilder.addDependency( driverServiceName, DriverService.ProviderClass.class, dataSourceService.getDriverServiceInjector() );
+
+            // Define JTA dependencies (required for XA)
+            serviceBuilder.addDependency( TxnServices.JBOSS_TXN_TRANSACTION_MANAGER, TransactionManager.class, dataSourceService.getTransactionManagerInjector() );
```
the ServiceBuilder API is fluent, you can chain the calls to `addDependency(...)`

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/parser/AgroalSubsystemParser_1_0.java:
```diff
@@ -0,0 +1,71 @@
+    static {
+        PersistentResourceXMLBuilder subsystemXMLBuilder = builder( AgroalSubsystemDefinition.INSTANCE.getPathElement(), Namespace.AGROAL_1_0.getUriString() );
+
+        PersistentResourceXMLBuilder datasourceXMLBuilder = builder( DataSourceDefinition.INSTANCE.getPathElement() );
```
-1

Attributes must be explicitly listed in the parser and not rely on their presence and order in the definition class.

The reason will manifest itself when you add a new attribute in the definition for the 2.0 version and make that attribute *also available in the 1.0 parser* without noticing it.

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/parser/AgroalSubsystemParser_1_0.java:
```diff
@@ -0,0 +1,71 @@
+        DriverDefinition.INSTANCE.getAttributes().forEach( driverXMLBuilder::addAttribute );
+        subsystemXMLBuilder.addChild( driverXMLBuilder );
+
+        XML_DESCRIPTION = subsystemXMLBuilder.build();
```
the whole XML builder is fluent and can be chained in a single statement (e.g. https://github.com/wildfly/wildfly/blob/f4b6b3afdea148862a28b34082c8dd215fa77607/undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemParser_6_0.java#L55)

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceDefinitionService.java:
```diff
@@ -0,0 +1,114 @@
+
+        try {
+            agroalDataSource = AgroalDataSource.from( dataSourceConfiguration, new LoggingDataSourceListener( dataSourceName ) );
+            AgroalLogger.SERVICE_LOGGER.infof( "Started datasource '%s' bound to [%s]", dataSourceName, jndiBinding );
```
info message must be i18ned

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceDefinitionService.java:
```diff
@@ -0,0 +1,114 @@
+            AgroalLogger.SERVICE_LOGGER.infof( "Started datasource '%s' bound to [%s]", dataSourceName, jndiBinding );
+        } catch ( SQLException e ) {
+            agroalDataSource = null;
+            throw new StartException( "Exception starting datasource " + dataSourceName, e );
```
exception must be i18ned

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceDefinitionService.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    public void stop(StopContext context) {
+        agroalDataSource.close();
+        AgroalLogger.SERVICE_LOGGER.infof( "Stopped datasource '%s'", dataSourceName );
```
info message must be i18ned

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+        Class<?> providerClass = driverService.getValue().providerClass();
+        if ( xa ) {
+            if ( !XADataSource.class.isAssignableFrom( providerClass ) ) {
+                throw new StartException( "An xa-datasource requires a javax.sqlXADataSource as connection provider. Fix the connection-provider for the driver" );
```
exception must be i18ned

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+            }
+        } else {
+            if ( providerClass != null && !DataSource.class.isAssignableFrom( providerClass ) && !Driver.class.isAssignableFrom( providerClass ) ) {
+                throw new StartException( "Invalid connection provider. Either a java.sql.Driver or javax.sql.DataSource implementation is required. Fix the connection-provider for the driver" );
```
idem

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+            NarayanaTransactionIntegration txIntegration = new NarayanaTransactionIntegration( transactionManager.getValue(), transactionSynchronizationRegistry.getValue(), jndiName, connectable );
+            dataSourceConfiguration.connectionPoolConfiguration( cp -> cp.transactionIntegration( txIntegration ) );
+        } else if ( xa ) {
+            throw new StartException( "Could not start xa-datasource: transaction manager is missing" );
```
idem

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+                    .addDependency( bindInfo.getParentContextServiceName(), ServiceBasedNamingStore.class, binderService.getNamingStoreInjector() )
+                    .install();
+
+            AgroalLogger.SERVICE_LOGGER.infof( "Started %sdatasource '%s' bound to [%s]", xa ? "xa-" : "", dataSourceName, jndiName );
```
idem

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+            AgroalLogger.SERVICE_LOGGER.infof( "Started %sdatasource '%s' bound to [%s]", xa ? "xa-" : "", dataSourceName, jndiName );
+        } catch ( SQLException e ) {
+            agroalDataSource = null;
+            throw new StartException( "Exception starting datasource " + dataSourceName, e );
```
ok, I'll mention it one last time :)  Any exception or log above info must be i18ned

jmesnil commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DriverService.java:
```diff
@@ -0,0 +1,69 @@
+        return providerClass;
+    }
+
+    public static final class ProviderClass {
```
what is the need for this class? why can't the DriverService directly provides the providerClass instead?

jmesnil commented on agroal/src/main/resources/org/wildfly/extension/agroal/LocalDescriptions.properties:
```diff
@@ -0,0 +1,106 @@
+agroal.flush-all=Flush all connections from the pool immediately
+agroal.flush-graceful=Flush all connections from the pool, allowing connections to be closed normally
+agroal.flush-invalid=Flush invalid connections from the pool
+agroal.flush-idle=Flush connections not being use
```
typo

jmesnil commented on agroal/src/main/resources/org/wildfly/extension/agroal/LocalDescriptions.properties:
```diff
@@ -0,0 +1,106 @@
+agroal.datasource.connection-pool.max-size=Maximum number of connections in the pool
+agroal.datasource.connection-pool.min-size=Minimum number of connections the pool should hold
+agroal.datasource.connection-pool.initial-size=Initial number of connections the pool should hold
+agroal.datasource.connection-pool.blocking-timeout-millis=Maximum time in milliseconds to block while waiting for a connection before throwing an exception
```
you don't need to specify the measurement unit as it is included in the attribute metadata

jmesnil commented on agroal/src/main/resources/schema/wildfly-agroal_1_0.xsd:
```diff
@@ -0,0 +1,324 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I will not review the XML schema as it should be derived from the management model instead of tweaking the parser/marshaller in `AgroalObjectAttribute`.

jmesnil commented on agroal/src/test/java/org/wildfly/extension/agroal/parser/SubsystemFullParsingTestCase.java:
```diff
@@ -0,0 +1,68 @@
+        ModelNode model = services.readWholeModel();
+        Assert.assertTrue( model.get( SUBSYSTEM ).hasDefined( AgroalExtension.SUBSYSTEM_NAME ) );
+
+        System.out.println( model );
```
Please remove the sysout from the tests.

jmesnil commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/agroal/main/module.xml:
```diff
@@ -0,0 +1,55 @@
+    </properties>
+
+    <exports>
+        <exclude path="org/wildfly/extension/agroal/logging"/>
```
If you move the logging in a `_private` subpackage, you will not need to exclude it from the module definition

jmesnil reviewed (changes requested):
Thanks for this PR, Agroal will become a key subsystem of WildFly and we need to put it in the best shape moving forward.

After reviewing this PR, I have several issues with it:

* the use of a custom AgroalObjectAttribute instead of regular ComplextObjectTypeAttribute (and the custom XML schema that results of this tweaking)
* the Parser must not use collections from the resource definition to list the attributes that can be parsed
* Exception and messages above info are not internationalized
* there are no docs in the admin guide for the subsystem
* there are no test validating the integration in the test suite. Maybe we could re-run some of the existing JDBC tests by replacing the `datasource` subsystem with `agroal`?

I've nitpicked other issues but they are not as important as the ones above.

thanks!

stuartwdouglas commented on agroal/src/main/java/org/wildfly/extension/agroal/parser/AgroalSubsystemParser_1_0.java:
```diff
@@ -0,0 +1,71 @@
+        PersistentResourceXMLBuilder subsystemXMLBuilder = builder( AgroalSubsystemDefinition.INSTANCE.getPathElement(), Namespace.AGROAL_1_0.getUriString() );
+
+        PersistentResourceXMLBuilder datasourceXMLBuilder = builder( DataSourceDefinition.INSTANCE.getPathElement() );
+        DataSourceDefinition.INSTANCE.getAttributes().forEach( datasourceXMLBuilder::addAttribute );
```
We should not be using method references, as they chew up metaspace, just use a loop (although as Jeff says you should not actually be looping like this in this case, they should be explicitly listed).

stuartwdouglas commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceDefinitionService.java:
```diff
@@ -0,0 +1,114 @@
+    public void start(StartContext context) throws StartException {
+        if ( transactionManager.getOptionalValue() != null && transactionSynchronizationRegistry.getOptionalValue() != null ) {
+            NarayanaTransactionIntegration txIntegration = new NarayanaTransactionIntegration( transactionManager.getValue(), transactionSynchronizationRegistry.getValue(), jndiBinding, false );
+            dataSourceConfiguration.connectionPoolConfiguration( cp -> cp.transactionIntegration( txIntegration ) );
```
We should not be using lambda's, they chew up too much metaspace.

stuartwdouglas commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+
+        if ( transactionManager.getOptionalValue() != null && transactionSynchronizationRegistry.getOptionalValue() != null ) {
+            NarayanaTransactionIntegration txIntegration = new NarayanaTransactionIntegration( transactionManager.getValue(), transactionSynchronizationRegistry.getValue(), jndiName, connectable );
+            dataSourceConfiguration.connectionPoolConfiguration( cp -> cp.transactionIntegration( txIntegration ) );
```
lambda

stuartwdouglas commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DataSourceService.java:
```diff
@@ -0,0 +1,136 @@
+            }
+        }
+
+        dataSourceConfiguration.connectionPoolConfiguration( cp -> cp.connectionFactoryConfiguration( cf -> cf.connectionProviderClass( providerClass ) ) );
```
lambdas

stuartwdouglas commented on agroal/src/main/java/org/wildfly/extension/agroal/service/DriverService.java:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+public class DriverService implements Service<DriverService.ProviderClass> {
```
Does this class need to exist? Why not just use ValueService?

stuartwdouglas commented on pom.xml:
```diff
@@ -216,6 +217,7 @@
         <version.gnu.getopt>1.0.13</version.gnu.getopt>
         <version.groovy-all>2.4.7</version.groovy-all>
         <version.httpunit>1.7.2</version.httpunit>
+        <version.io.agroal>0.5</version.io.agroal>
```
This version should really be 1.0 for inclusion

dmlloyd commented on agroal/src/main/java/org/wildfly/extension/agroal/deployment/DataSourceDefinitionInjectionSource.java:
```diff
@@ -0,0 +1,238 @@
+ */
+public class DataSourceDefinitionInjectionSource extends ResourceDefinitionInjectionSource {
+
+    public static final ServiceName DATASOURCE_DEFINITION_SERVICE_PREFIX = ServiceName.of( "wildfly", "agroal", "datasource-definition" );
```
In principle it's OK, but since there is no ``"wildfly"`` shared constant I think we should stick with ``ServiceName.JBOSS`` as a root as a purely practical matter.

jmesnil commented:
@barreiro sorry, there were 2 other things I forgot to mention in my review.

* Could you rename the subsytem `datasources-agroal` instead of `agroal`? We have the convention for new subsystems to name them `<feature>-<provider>` (e.g. `messaging-activemq`). This helps users have a better idea of what the subsytem provides.
* Could you also add an standalone configuration example in https://github.com/wildfly/wildfly/tree/master/feature-pack/src/main/resources/configuration/examples? That'd be awesome to have a standalone-agroal.xml  similar to standalone-full.xml except that the default EE7 `DataSource` is provided by the agroal subsystem instead of the current one.

jamezp commented:
The Galleon PR has been merged. Dependencies added to the `feature-pack/pom.xml` will also need to be added to the `galleon-pack/pom.xml`.

rhusar commented on agroal/src/main/java/org/wildfly/extension/agroal/definition/AbstractDataSourceDefinition.java:
```diff
@@ -0,0 +1,253 @@
+    public static final int LEAK_DETECTION_DEFAULT_VALUE = 0;
+    public static final int IDLE_REMOVAL_DEFAULT_VALUE = 0;
+
+    public static final SimpleAttributeDefinition JNDI_NAME_ATTRIBUTE = create( "jndi-name", ModelType.STRING )
```
@barreiro Really the whole thing needs a reformat with WF style :-) https://github.com/wildfly/wildfly-core/tree/master/ide-configs should be up to date.

jamezp commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Datasources_Agroal.adoc:
```diff
@@ -0,0 +1,149 @@
+    <drivers>
+        <driver name="h2" module="com.h2database.h2" class="org.h2.Driver"/>
+    </drivers>
+</subsystem>
```
We should be using CLI examples instead of XML. Users are suggested to not edit the raw XML.

stuartwdouglas commented:
retest this please

simkam commented on datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/service/DataSourceDefinitionService.java:
```diff
@@ -0,0 +1,114 @@
+import org.jboss.as.naming.ManagedReferenceFactory;
+import org.jboss.as.naming.deployment.ContextNames;
+import org.jboss.msc.inject.Injector;
+import org.jboss.msc.service.Service;
```
I wonder if it wouldn't be better to update it to new Service API, see. https://github.com/wildfly/wildfly-proposals/blob/master/msc/WFLY-10254_WFCORE-3752_MSC_Upgrade.adoc

simkam commented on feature-pack/src/main/resources/modules/system/layers/base/io/agroal/main/module.xml:
```diff
@@ -0,0 +1,41 @@
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="io.agroal">
```
urn:jboss:module:1.8 ?

simkam commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/datasources-agroal/main/module.xml:
```diff
@@ -0,0 +1,51 @@
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="org.wildfly.extension.datasources-agroal">
```
urn:jboss:module:1.8 ?

darranl reviewed (changes requested):
I notice all of the classes added have public visibility by default - is that strictly required?

Generally we should drop to default visibility except for the cases we know public is needed for a reason.

Even if a module is flagged private where classes are public there is a risk of cross subsystem dependencies getting added, this will create further problems later when provisioning may or may not add the modules of a specific subsystem.

jamezp commented:
The following change will need to be added so the build will work https://github.com/jamezp/wildfly/commit/855fc4184359d9be32940459acbcbccc6b81c30f.

There is also a lack of integration tests to prove this actually works. Trying it out with the batch subsystem trying to use a `jdbc-job-repository` showed the datasource resource does not register the `org.wildfly.data-source` capability.

I'm curious too if any thought has been given to whether or not the `*-ds.xml` descriptors will be supported or at least partially supported. There was quite a bit of push back from the community when https://issues.jboss.org/browse/WFLY-4296 was filed.

barreiro commented:
Updated the PR. Main change has been moving classes to same package and drop their visibility. Added the subsystem to the galleon build. Added datasource capabilities.

@simkam I have updated the main services to the new API (there are places where the old API is used that I can't work around)

dmlloyd commented:
IMO it is OK to use the old API as well for now.  I think the new API will not be _required_ for new code until we solve a couple remaining issues in management-MSC integration.

rhusar commented:
> until we solve a couple remaining issues in management-MSC integration.

@dmlloyd Can you please elaborate or point to the Jira/discussion what changes these are?

dmlloyd commented:
There aren't any discussions yet; it's something I'm still trying to formulate.  The gist of the issue is that a lot of our management operations count on being able to directly access service objects in order to set run time parameters, but this doesn't always work well (or at all) when a service provides multiple values.

simkam commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Datasources_Agroal.adoc:
```diff
@@ -0,0 +1,238 @@
+
+[source, ruby]
+----
+[standalone@localhost:9990  /] /subsystem=datasources-agroal/datasource=sample:read-resource(include-runteime)
```
runtime

simkam commented on datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/AbstractDataSourceDefinition.java:
```diff
@@ -0,0 +1,247 @@
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition PASSWORD_ATTRIBUTE = create("password", ModelType.STRING)
```
I wonder if it shouldn't support also credential-reference, users don't like clear text password in configuration.

barreiro replied:
yes! it's something on the top of my priority list.

darranl replied:
Has the credential-reference been added?

barreiro replied:
@darranl yes!

simkam commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Datasources_Agroal.adoc:
```diff
@@ -0,0 +1,238 @@
+
+The Agroal subsystem allows the definition of datasources. Declaring a new datasource consists of two separate steps: provide a JDBC driver and define a datasource that references the driver you installed.
+
+The Agroal subsystem is provided by the https://agroal.github.io/[Agroal] project. For a detailed description of the available configuration properties, please consult the project documentation.
```
is there such documentation? I didn't find it.

barreiro replied:
I'm working on it!

simkam commented on datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/AbstractDataSourceDefinition.java:
```diff
@@ -0,0 +1,247 @@
+            .setRestartAllServices()
+            .build();
+
+    static final ObjectTypeAttributeDefinition CONNECTION_FACTORY_ATTRIBUTE = ObjectTypeAttributeDefinition.create("connection-factory", DRIVER_ATTRIBUTE, URL_ATTRIBUTE, TRANSACTION_ISOLATION_ATTRIBUTE, NEW_CONNECTION_SQL_ATTRIBUTE, USERNAME_ATTRIBUTE, PASSWORD_ATTRIBUTE, CONNECTION_PROPERTIES_ATTRIBUTE)
```
Are we sure we want to have "url" as common attribute for connection factory for all, driver, datasource and xa-datasource? I see in agroal that if driver class is javax.sql.Datasource or javax.sql.XADatasource, "url" is injected using "setUrl" or "setProperty". I'm not sure if it's good idea. There are drivers that haven't such setters and use "serverName", "port" and "databaseName" instead of "url". In such case agroal throws NoSuchMethodException which ends as WARN in WF log.

barreiro replied:
ok, I'll look into this.

simkam commented on datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/DriverDefinition.java:
```diff
@@ -0,0 +1,87 @@
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition CLASS_ATTRIBUTE = create("class", ModelType.STRING)
```
wouldn't be better to add more attributes for different types? something like
driver-class - if not specified, it is detected from /services/java.sql.Driver
datasource-class
xa-datasource-class

where datasource would have priority over driver

it's easy now to end with invalid configuration. xa-datasource requires javax.sql.XADataSource

simkam commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Datasources_Agroal.adoc:
```diff
@@ -0,0 +1,238 @@
+[[agroal]]
+= Agroal configuration
```
Can you add also note that agroal doesn't support lazy enlistment

barreiro replied:
Sure! Thanks for pointing that out.

fmarchioni commented:
I'm setting up my environment to run transaction tests with datasources-agroal. One thing I' ve noticed there is no test-connection CLI command available:
```
/subsystem=datasources-agroal/xa-datasource=TestDatasource:
add                         list-add                    map-get                     read-attribute-group        read-operation-description  reset-statistics  
```
. . . . 
is there an equivalent way to test if the Connections are valid?

barreiro commented:
Added initial elytron support (credential-reference and authentication-context). Added integration tests. Added test-connection operation. Updated to Agroal 1.1 to fix an issue with Sybase drivers. Fix misc typos in docs.

msve commented:
@barreiro integration with Elytron is part of different RFE (EAP7-1083) which is planned to EAP 7.3. It should be rolled back, we don't want multiple RFEs in one PR.

Please leave this PR for the scope of EAP7-983 only. If you want early feedback for EAP7-1083, open it as separate followup PR.

darranl commented:
Yeah is is mentioned in the analysis linked on this PR - I think it is important to be in this PR so a new resource being added that needs a password has credential store integration from the outset.

barreiro commented:
@msve technically is possible, but the point here is to have the required resources in place right away (reference to credential storage and authentication context), while while the testing is being taken care of.
Like @darranl pointed out, elytron integration is mentioned in the requirements.

barreiro commented:
Revert to back to use old Service API to work around WFCORE-3982

jamezp commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/agroal/AgroalDatasourceTestBase.java:
```diff
@@ -0,0 +1,295 @@
+            addSystemProperty(jndiName, defaultPropertyDs.getJndiName());
+            addSystemProperty(driverModule, defaultPropertyDs.getDriverModule());
+            addSystemProperty(driverClass, defaultPropertyDs.getDriverClass());
+            
```
The spaces from this line need to be removed as it's causing a checkstyle issue.

bstansberry commented:
I'm ok with merging this for 14.0.0.Beta2 but there are a number of things that must be addressed for CR1,  one that perhaps must be addressed but if not is a Critical for 15, and then some misc points. Instead of making a bunch of line comments, I'll write it all here. Please someone file several JIRAs out of this but I'm just going to use this comment. Please ping @yersan @jmesnil or myself for assistance with these. Myself as last choice. ;) Lots of other people can be helpful too but the 3 of us do management code as a primary task.

Blockers for WF 14 CR1:

1) The DS/XADS 'statistics' attribute is not properly described. It has a static set of response fields and the value of each field has a known type, nillability etc.

2) The DS/XADS 'statistics' attribute should not be registered in an HC process. Use ManagementResourceRegistration.getProcessType().isServer() to determine if the process is a server.

3) The driver class-info attribute should not be registered in an HC process. Use ManagementResourceRegistration.getProcessType().isServer() to determine if the process is a server.

4) The DS/XADS runtime-only operations (flush etc) should not be registered in an HC process. Use ManagementResourceRegistration.getProcessType().isServer() to determine if the process is a server. See https://developer.jboss.org/docs/DOC-55978 for conceptual information about supporting ops like this in the /profile=* resources. Perhaps some of them could be supported following further analysis and testing, but in this first iteration I don't want them supported.

5) The 'module' and 'classname' attributes on the driver resource allows expressions.  A bit of black art is we don't allow expressions on those kinds of attributes.  TBH there's no great reason for it, but we should be consistent.

6) The class-info attribute on the driver is not properly described. It's a map of property name -> single param name, not a primitive list.

7) DataSourceDefinitionDescriptorProcessor is missing i18n in some exception messages

8) AgroalSubsystemAdd.populateModel should be removed

9) AgroalSubsystemRemove should be removed; use ReloadRequiredRemoveStepHandler

10) AIUI the only correct source for a DS/XADS driver is one of the driver resources. So, those drivers should expose a capability (should be private) and the driver attribute on the DS/XADS should be a capability reference.

11) The service name used for the DS/XADS is not correct. It needs to be driven by the DATASOURCE_RUNTIME_CAPABILITY, otherwise other subsystems cannot use the capability to access the service.

Criticals:

1) We need to add RBAC application type constraints for these resources so admins in the Deployer RBAC role can be allowed to administer them. Now they cannot which is different from the IJ-based subsystem. See for example https://github.com/wildfly/wildfly/blob/master/connector/src/main/java/org/jboss/as/connector/subsystems/datasources/DataSourceDefinition.java#L84 

I'm inclined to say we defer this to the next release though as it is late to test this and being overly restrictive is safer than being looser but poorly tested.

Other points:

These aren't critical, they are more FYI stuff, polish bits etc.

1) There's a lot of if (XXX.resolveModelAttribute(...).isDefined()) { sometype xxx = XXX.resolveModelAttribute(...).asSomeType(); ... }
That results in duplicate resolution. Instead, DMR now lets you:

sometype xxx = XXX.resolveModelAttribute(...).asSomeTypeOrNull();
if (xxx != null) {...}

2) In some of the attribute definitions there are things like new ModelNode(0) which can be ModelNode.ZERO, ModelNode.TRUE, ModelNode.FALSE etc. Also newish in DMR.

3) There's a fair amount of 'protected' in AbstractDataSourceOperations which itself is package protected. Perhaps similar things elsewhere. I hate 'protected' so prefer that it never appear unless strictly necessary. :) Experience has taught me it's a maintenance burden waiting to happen.

4) XADataSourceAdd : String datasourceName = PathAddress.pathAddress(operation.require(OP_ADDR)).getLastElement().getValue();
That should be String datasourceName = context.getCurrentAddressValue();

rsvoboda commented:
@barreiro 
> Added initial elytron support (credential-reference and authentication-context).

What are the remaining pieces for elytron support ? This should be mentioned in analysis document as not in the scope for TP.

simkam commented:
I created jiras for items from Brian's review
https://issues.jboss.org/browse/WFLY-10866 for 1,6
https://issues.jboss.org/browse/WFLY-10868 for 5
https://issues.jboss.org/browse/WFLY-10868 for 7
https://issues.jboss.org/browse/WFLY-10869 for 8,9
https://issues.jboss.org/browse/WFLY-10870 for 10 
https://issues.jboss.org/browse/WFLY-10871 for 11

rsvoboda commented:
I think those issues should be Blockers as they needs to be done or Agroal will be pulled away.

barreiro commented:
@rsvoboda elytron integration needs proper testing. Also the Agroal library needs to implement support for more credential types (supplied by the authentication-context).
